File: CleverFerretV2/core/network/src/main/java/com/cleverferret/v2/core/network/resilience/IdempotentRetryPolicy.java
package com.cleverferret.v2.core.network.resilience;

import com.cleverferret.v2.core.common.error.AppError;

public final class IdempotentRetryPolicy implements RetryPolicy {
    private final int maxAttempts;
    private final long baseDelayMillis;

    public IdempotentRetryPolicy() {
        this(4, 200L);
    }

    public IdempotentRetryPolicy(int maxAttempts, long baseDelayMillis) {
        this.maxAttempts = maxAttempts;
        this.baseDelayMillis = baseDelayMillis;
    }

    @Override
    public boolean shouldRetry(AppError error, int attempt) {
        if (attempt >= maxAttempts) {
            return false;
        }
        return error instanceof AppError.TransportFailure
                || error instanceof AppError.RateLimitFailure
                || error instanceof AppError.ExternalServiceFailure;
    }

    @Override
    public long delayMillis(int attempt) {
        return (long) (baseDelayMillis * Math.pow(2, Math.max(0, attempt - 1)));
    }

    @Override
    public int maxAttempts() {
        return maxAttempts;
    }
}

File: CleverFerretV2/core/network/src/main/java/com/cleverferret/v2/core/network/resilience/IdempotentRetryPolicy.kt
package com.cleverferret.v2.core.network.resilience

import com.cleverferret.v2.core.common.error.AppError
import kotlin.math.max
import kotlin.math.pow

class IdempotentRetryPolicy(
    private val retryMaxAttempts: Int = 4,
    private val baseDelayMillis: Long = 200L,
) : RetryPolicy {
    override fun shouldRetry(error: AppError, attempt: Int): Boolean =
        attempt < retryMaxAttempts && (error is AppError.TransportFailure || error is AppError.RateLimitFailure || error is AppError.ExternalServiceFailure)

    override fun delayMillis(attempt: Int): Long = (baseDelayMillis * 2.0.pow(max(0, attempt - 1))).toLong()
    override fun maxAttempts(): Int = retryMaxAttempts
}
